combine/combine.go

```diff
@@ -0,0 +1,514 @@
+)
+
+const bytesInMB = 1000000
+const pgzipWriterBlocksMultiplier = 2
+const combineStatsOutputFileBasename = "combine.stats.gz"
+const combineUserGroupOutputFileBasename = "combine.byusergroup.gz"
+const combineGroupOutputFileBasename = "combine.bygroup"
+const combineDGUTOutputFileBasename = "combine.dgut.db"
+const combineLogOutputFileBasename = "combine.log.gz"
+const numSummaryColumns = 2
+const numSummaryColumnsDGUT = 3
+const groupSumCols = 2
+const userGroupSumCols = 3
+const intBase = 10
+const dgutSumCols = 4
+const dgutStoreBatchSize = 100000
+
+const statLogOutputFileSuffix = ".log"
+const statOutputFileSuffix = ".stats"
+const statDGUTSummaryOutputFileSuffix = ".dgut"
+const statGroupSummaryOutputFileSuffix = ".bygroup"
+const statUserGroupSummaryOutputFileSuffix = ".byusergroup"
+const userOnlyPerm = 448
+
+type Error string
+
+type Combine struct {
+	SourceDir string
+
+	Suffixes map[string]string
+
+	Functions [3]mergeStreamToOutputFunc
+}
+
+func (e Error) Error() string { return string(e) }
+
+func (c *Combine) Combine() error {
+	sourceDir, err := filepath.Abs(c.SourceDir)
+	if err != nil {
+		return err
+	}
+
+	// Merge and compress byusergroup files.
+	if err := MergeAndOptionallyCompressFiles(sourceDir, statUserGroupSummaryOutputFileSuffix,
+		combineUserGroupOutputFileBasename, mergeUserGroupStreamToCompressedFile); err != nil {
+		return err
+	}
+
+	// Merge and compress log files.
+	if err := MergeAndOptionallyCompressFiles(sourceDir, statLogOutputFileSuffix,
+		combineLogOutputFileBasename, mergeLogStreamToCompressedFile); err != nil {
+		return err
+	}
+
+	// Merge bygroup files.
+	if err := MergeAndOptionallyCompressFiles(sourceDir, statGroupSummaryOutputFileSuffix,
+		combineGroupOutputFileBasename, mergeGroupStreamToFile); err != nil {
+		return err
+	}
+
+	// Compress and concatenate stat files.
+	if err := CompressAndConcatenate(sourceDir, filepath.Join(sourceDir, combineStatsOutputFileBasename)); err != nil {
+		return err
+	}
+
+	// Merge dgut files.
+	if err := MergeDGUTFilesToDB(sourceDir); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// MergeAndOptionallyCompressFiles identifies files from the supplied suffix,
+// merges them, and then may or may not compress them based upon the type of
+// function supplied as mergeAndStreamFunc.
+func MergeAndOptionallyCompressFiles(sourceDir, inputFileSuffix, outputFileBasename string,
+	mergeAndStreamFunc mergeStreamToOutputFunc) error {
+	paths, err := findFilePathsInDir(sourceDir, inputFileSuffix)
+	if err != nil {
+		return err
+	}
+
+	output, err := createOutputFileInDir(sourceDir, outputFileBasename)
+	if err != nil {
+		return err
+	}
+
+	err = mergeFilesAndStreamToOutput(paths, output, mergeAndStreamFunc)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// findFilePathsInDir finds files in the given dir that have basenames with the
+// given suffix. Dies on error.
+func findFilePathsInDir(dir, suffix string) ([]string, error) {
```

Comment:
We're either not going to have a function like this (leaving it in cmd/combine.go), or we'll make it public in another new package that might be called 'fs'. The functions in this package will just take os.File or path strings as inputs.